R/join_keys.R

```diff
@@ -0,0 +1,697 @@
+  }
+
+  # Accepting 1 subscript with valid `value` formal
+  checkmate::assert_list(value, names = "named", types = "character", null.ok = TRUE)
+
+  join_keys_obj <- NextMethod("[[<-", jk)
+
+  # Keep original parameters as variables will be overwritten for `NextMethod` call
+  original_value <- value
+  ds1 <- dataset_1
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(value)) {
+    if (ds2 == ds1) next
+
+    value <- rlang::`%||%`(join_keys_obj[[ds2]], list())
+    new_value <- original_value[[ds2]]
+
+    if (
+      checkmate::test_character(new_value, min.len = 1) &&
+        is.null(names(new_value))
+    ) {
+      new_value <- setNames(new_value, new_value)
+    } else if (
+      checkmate::test_character(new_value, min.len = 1)
+    ) {
+      # Invert key
+      new_value <- setNames(names(new_value), new_value)
+    }
+
+    # Change variables for NextMethod call
+    dataset_1 <- ds2
+    value[[ds1]] <- new_value
+    join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+  }
+
+  join_keys_obj
+}
+
+#' Mutate `JoinKeys` with a new values
+#'
+#' @description `r lifecycle::badge("experimental")`
+#' Mutate `JoinKeys` with a new values
+#'
+#' @param x (`JoinKeys`) object to be modified
+#' @param dataset_1 (`character`) one dataset name
+#' @param dataset_2 (`character`) other dataset name
+#' @param value (named `character`) column names used to join
+#'
+#' @return modified `JoinKeys` object
+#'
+#' @export
+mutate_join_keys <- function(x, dataset_1, dataset_2, value) {
+  UseMethod("mutate_join_keys")
+}
+
+#' @rdname mutate_join_keys
+#' @export
+#' @examples
+#'
+#' # JoinKeys ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "some_col")
+#' mutate_join_keys(jk, "ds2", "ds3", "another")
+mutate_join_keys.JoinKeys <- function(x, dataset_1, dataset_2, value) {
+  checkmate::assert_string(dataset_1)
+  checkmate::assert_string(dataset_2)
+  checkmate::assert_character(value, any.missing = FALSE)
+
+  res <- join_pair(x, join_key(dataset_1, dataset_2, value))
+
+  logger::log_trace(
+    sprintf(
+      "JoinKeys updated the keys between %s and %s to %s",
+      dataset_1,
+      dataset_2,
+      paste(value, collapse = ", ")
+    )
+  )
+
+  res
+}
+
+#' @rdname mutate_join_keys
+#' @export
+#' @examples
+#'
+#' # teal_data ----
+#'
+#' ADSL <- teal.data::example_cdisc_data("ADSL")
+#' ADRS <- teal.data::example_cdisc_data("ADRS")
+#'
+#' x <- cdisc_data(
+#'   "ADSL" = ADSL,
+#'   "ADRS" = ADRS
+#' )
+#' join_keys(x)["ADSL", "ADRS"]
+#'
+#' join_keys(x) <- mutate_join_keys(x, "ADSL", "ADRS", c("COLUMN1" = "COLUMN2"))
+#' join_keys(x)["ADSL", "ADRS"]
+mutate_join_keys.teal_data <- function(x, dataset_1, dataset_2, value) { # nolint
+  join_keys(x) <- mutate_join_keys(join_keys(x), dataset_1, dataset_2, value)
+  join_keys(x)
+}
+
+#' @rdname split_join_keys
+#' @export
+split_join_keys <- function(join_keys_obj) {
+  UseMethod("split_join_keys", join_keys_obj)
+}
```

Comment:
Existed before, but I just checked within `insightsengineering` and it seems it's only used on our tests (`JoinKeys$split()`)

It's a nice functionality to have (deconstructor), but given it's not used once I think it can be safely removed.

Reply:
Done in 901bcec7